coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/CoralSqlNodeToTrinoSqlNodeConverter.java

```diff
@@ -0,0 +1,235 @@
+ */
+package com.linkedin.coral.trino.rel2trino;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFieldImpl;
+import org.apache.calcite.rel.type.RelRecordType;
+import org.apache.calcite.sql.JoinConditionType;
+import org.apache.calcite.sql.JoinType;
+import org.apache.calcite.sql.SqlBasicCall;
+import org.apache.calcite.sql.SqlBasicTypeNameSpec;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCharStringLiteral;
+import org.apache.calcite.sql.SqlDataTypeSpec;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlJoin;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.SqlTypeNameSpec;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.util.SqlShuttle;
+import org.apache.calcite.sql.validate.SqlValidatorUtil;
+
+import com.linkedin.coral.com.google.common.collect.ImmutableList;
+import com.linkedin.coral.common.functions.CoralSqlUnnestOperator;
+import com.linkedin.coral.trino.rel2trino.functions.TrinoArrayTransformFunction;
+
+import static org.apache.calcite.rel.rel2sql.SqlImplementor.*;
+import static org.apache.calcite.sql.parser.SqlParserPos.*;
+
+
+/**
+ * CoralSqlNodeToTrinoSqlNodeConverter rewrites the Coral SqlNode AST. It replaces Coral IR SqlCalls
+ * with Trino compatible SqlCalls to subsequently obtain a Trino compatible SqlNode AST representation.
+ * This will enable generating a SQL which can be accurately interpreted by the Trino engine.
+ *
+ * This is achieved by visiting the Coral SqlNode AST in a pre-order traversal manner and
+ * transforming each SqlNode (SqlCall), wherever required.
+ * The transformation may involve change in operator, reordering the operands
+ * or even re-constructing the SqlCall.
+ */
+public class CoralSqlNodeToTrinoSqlNodeConverter extends SqlShuttle {
+
+  public CoralSqlNodeToTrinoSqlNodeConverter() {
+  }
+
+  @Override
+  public SqlNode visit(final SqlCall call) {
+    SqlCall transformedSqlCall = getTransformedSqlCall(call);
+    return super.visit(transformedSqlCall);
+  }
+
+  public static SqlCall getTransformedSqlCall(SqlCall sqlCall) {
+    switch (sqlCall.getOperator().kind) {
+      case SELECT:
+        return getTransformedSqlSelectSqlCall(sqlCall);
+      case JOIN:
+        return getTransformedJoinSqlCall(sqlCall);
+      case AS:
+        return getTransformedAsSqlCall(sqlCall);
+      case UNNEST:
+        return getTransformedUnnestSqlCall(sqlCall);
+      case EQUALS:
+        return getTransformedEqualsOperatorSqlCall(sqlCall);
```

Comment:
Could you add more descriptive comments for all the transformation methods (i.e. why we need this transformation), preferably with examples? It's kind of hard to understand these methods just by reading code.